File: mobile/android/focus-android/app/src/testWebkit/java/org/mozilla/focus/webview/matcher/TrieTest.java
package org.mozilla.focus.webview.matcher;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mozilla.focus.webview.matcher.Trie.WhiteListTrie;
import org.mozilla.focus.webview.matcher.util.FocusString;
import org.robolectric.RobolectricTestRunner;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;

@RunWith(RobolectricTestRunner.class)
public class TrieTest {

    @Test
    public void findNode() throws Exception {
        final Trie trie = Trie.createRootNode();

        assertNull(trie.findNode(FocusString.create("hello")));

        final Trie putNode = trie.put(FocusString.create("hello"));
        final Trie foundNode = trie.findNode(FocusString.create("hello"));

        assertNotNull(putNode);
        assertNotNull(foundNode);
        assertEquals(putNode, foundNode);

        // Substring matching: doesn't happen (except for subdomains, we test those later)
        assertNull(trie.findNode(FocusString.create("hell")));
        assertNull(trie.findNode(FocusString.create("hellop")));

        trie.put(FocusString.create("hellohello"));

        // Ensure both old and new overlapping strings can still be found
        assertNotNull(trie.findNode(FocusString.create("hello")));
        assertNotNull(trie.findNode(FocusString.create("hellohello")));

        // These still don't match:
        assertNull(trie.findNode(FocusString.create("hell")));
        assertNull(trie.findNode(FocusString.create("hellop")));

        // Domain specific / partial domain tests:
        trie.put(FocusString.create("foo.com").reverse());

        // Domain and subdomain can be found
        assertNotNull(trie.findNode(FocusString.create("foo.com").reverse()));
        assertNotNull(trie.findNode(FocusString.create("bar.foo.com").reverse()));
        // But other domains with some overlap don't match
        assertNull(trie.findNode(FocusString.create("bar-foo.com").reverse()));
        assertNull(trie.findNode(FocusString.create("oo.com").reverse()));
    }

    @Test
    public void testWhiteListTrie() {
        final WhiteListTrie trie;

        {
            final Trie whitelist = Trie.createRootNode();

            whitelist.put(FocusString.create("abc"));

            trie = WhiteListTrie.createRootNode();
            trie.putWhiteList(FocusString.create("def"), whitelist);
        }

        assertNull(trie.findNode(FocusString.create("abc")));

        // In practice EntityList uses it's own search in order to cover all possible matching notes
        // (e.g. in case we have separate whitelists for mozilla.org and foo.mozilla.org), however
        // we don't need to test that here yet.
        final WhiteListTrie foundWhitelist = (WhiteListTrie) trie.findNode(FocusString.create("def"));
        assertNotNull(foundWhitelist);

        assertNotNull(foundWhitelist.whitelist.findNode(FocusString.create("abc")));
    }
}

File: mobile/android/android-components/components/browser/engine-system/src/test/java/mozilla/components/browser/engine/system/matcher/TrieTest.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package mozilla.components.browser.engine.system.matcher

import androidx.test.ext.junit.runners.AndroidJUnit4
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class TrieTest {

    @Test
    fun findNode() {
        val trie = Trie.createRootNode()

        assertNull(trie.findNode("hello"))
        val putNode = trie.put("hello")
        val foundNode = trie.findNode("hello")
        assertNotNull(putNode)
        assertNotNull(foundNode)
        assertEquals(putNode, foundNode)

        // Substring matching doesn't happen (except for subdomains)
        assertNull(trie.findNode("hell"))
        assertNull(trie.findNode("hellop"))

        // Ensure both old and new overlapping strings can still be found
        trie.put("hellohello")
        assertNotNull(trie.findNode("hello"))
        assertNotNull(trie.findNode("hellohello"))
        assertNull(trie.findNode("hell"))
        assertNull(trie.findNode("hellop"))

        // Domain and subdomain can be found
        trie.put("foo.com".reverse())
        assertNotNull(trie.findNode("foo.com".reverse()))
        assertNotNull(trie.findNode("bar.foo.com".reverse()))
        assertNull(trie.findNode("bar-foo.com".reverse()))
        assertNull(trie.findNode("oo.com".reverse()))
    }
}
